rfd/0186-optionally-require-reason-for-access-request.md

```diff
@@ -0,0 +1,215 @@
+affected.
+
+Example:
+
+For two roles `rr-role` and `dba-req-role` as defined below:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: rr-role
+spec:
+  options:
+    request_access: reason-required
+```
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: dba-req-role
+spec:
+  allow:
+    request:
+      roles:
+        - 'dba'
+```
+
+A user with both roles `rr-role` and `dba-req-role` assigned will be affected
+by request_access set in role `rr-role` while requesting access to role `dba`
+allowed by role `dba-req-role`.
+
+Because request_access option is global a priority has to be establish when
+more than one role has the option specified:
+
+- reason
+- always
+- reason-required
+- optional (default)
+
+Problems:
+
+- It may be confusing/undesired that roles with request_access option set to
+  "reason-required" affects other roles.
+
+
+#### 2. Add a new value to role.options.request_access and change the option to be scoped
+
+> [!NOTE]
+> This is a breaking change.
+
+The solution would be like in 1. but with changed semantics of request_access
+option. It would become role-scoped instead of global.
+
+Example:
+
+For two roles `dba-requester` and `ssh-requester` as defined below:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: ssh-requester
+spec:
+  allow:
+    request:
+      roles:
+        - 'ssh'
+  options:
+    request_access: reason-required
+```
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: dba-requester
+spec:
+  allow:
+    request:
+      roles:
+        - 'dba'
+  options:
+    request_access: optional # default, doesn't need to be set explicitly
+```
+
+A user with `ssh-requester` and `dba-requester` roles assigned would have to
+provide a reason while creating an access request for role `ssh` but wouldn't
+have to provide reason while requesting access to `dba`.
+
+The example above is clear and simple, but there are problems with the
+implementation:
+
+- **This is a breaking change.** It may come as a surprise to the existing
+  users that only _some_ instead of _all_ roles are automatically requested on
+  login when request_access is set to "always" or "reason".
+- What happens when there are 2 or more roles allowing requesting the same
+  role/resource but they have different request_access set? Should the solution
+  fall back to priorities, like in 1.?
+
+
+#### 3. Add a new role.spec.allow.request.reason_required setting
+
+This is pretty self-explanatory. role.spec.allow.reason_required could be set
+to true/false and being false by default.
+
+Example:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: dba-requester
+spec:
+  allow:
+    request:
+      require_reason: true
+      roles:
+        - 'dba'
+```
+
+
+The problems with the approach:
+
+- It isn't clear what should happen when when there is a role with
+  `options.request_access: always` and a role with
+  `allow.request.reason_required: true`. Should the reason be requested
+  regardless during login?
```

Comment:
yes, the reason should be required for all access requests, automatic or not. Mostly no one uses the automatic access requests